sdk/src/resource_store.rs

```diff
@@ -0,0 +1,211 @@
+// Copyright 2023 Adobe. All rights reserved.
+// This file is licensed to you under the Apache License,
+// Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+// or the MIT license (http://opensource.org/licenses/MIT),
+// at your option.
+
+// Unless required by applicable law or agreed to in writing,
+// this software is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR REPRESENTATIONS OF ANY KIND, either express or
+// implied. See the LICENSE-MIT and LICENSE-APACHE files for the
+// specific language governing permissions and limitations under
+// each license.
+
+#[cfg(feature = "file_io")]
+use std::path::{Path, PathBuf};
+use std::{borrow::Cow, collections::HashMap};
+
+use serde::{Deserialize, Serialize};
+use uuid_b64::UuidB64;
+
+use crate::{Error, Result};
+
+/// Function that is used by serde to determine whether or not we should serialize
+/// resources based on the `serialize_resources` flag.
+/// (Serialization is disabled by default.)
+pub(crate) fn skip_serializing_resources(_: &ResourceStore) -> bool {
+    !cfg!(feature = "serialize_thumbnails") || cfg!(test)
+}
+
+/// A reference to a resource to be used in JSON serialization
+#[derive(Clone, Debug, Deserialize, Serialize, Eq, PartialEq)]
+pub struct ResourceRef {
+    pub content_type: String,
+    pub identifier: String,
+}
+
+impl ResourceRef {
+    pub fn new<S: Into<String>>(content_type: S, identifier: S) -> Self {
+        Self {
+            content_type: content_type.into(),
+            identifier: identifier.into(),
+        }
+    }
+
+    pub fn from_content_type<S: Into<String>>(content_type: S) -> Self {
+        let content_type = content_type.into();
+        let identifier = ResourceStore::content_type_id(&content_type);
+        Self {
+            content_type,
+            identifier,
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct Config {}
+
+/// Resource store to contain binary objects referenced from JSON structures
+#[derive(Debug, Serialize)]
+pub struct ResourceStore {
+    resources: HashMap<String, Vec<u8>>,
+    #[cfg(feature = "file_io")]
+    base_path: Option<PathBuf>,
+}
+
+impl ResourceStore {
+    pub fn new() -> Self {
+        ResourceStore {
+            resources: HashMap::new(),
+            #[cfg(feature = "file_io")]
+            base_path: None,
+        }
+    }
+
+    #[cfg(feature = "file_io")]
+    pub fn base_path(&self) -> Option<&Path> {
+        self.base_path.as_deref()
+    }
+
+    #[cfg(feature = "file_io")]
+    pub fn set_base_path<P: Into<PathBuf>>(&mut self, base_path: P) {
+        self.base_path = Some(base_path.into());
+    }
+
+    ///  generate a unique id for a given content type (adds a file extension)
+    pub fn content_type_id(format: &str) -> String {
```

Comment:
We probably need to consolidate this logic someday since we now have a number of places to update when we add/delete supported formats.

Reply:
yes I'm planning on a PR soon to move all this into one place.